prow/cmd/build/controller.go

```diff
@@ -17,46 +17,424 @@ limitations under the License.
+
+		if err := reconcile(c, key.(string)); err != nil {
+			runtime.HandleError(fmt.Errorf("failed to reconcile %s: %v", key, err))
+			continue // Do not forget so we retry later.
+		}
+		c.workqueue.Forget(key)
+	}
+}
+
+// enqueueKey schedules an item for reconciliation.
+func (c *controller) enqueueKey(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.AddRateLimited(key)
+}
+
+type reconciler interface {
+	getProwJob(namespace, name string) (*prowjobv1.ProwJob, error)
+	getBuild(namespace, name string) (*buildv1alpha1.Build, error)
+	deleteBuild(namespace, name string) error
+	createBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error)
+	updateBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error)
+	updateProwJob(namespace string, pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error)
+	now() metav1.Time
+}
+
+func (c *controller) getProwJob(namespace, name string) (*prowjobv1.ProwJob, error) {
+	return c.pjLister.ProwJobs(namespace).Get(name)
+}
+
+func (c *controller) getBuild(namespace, name string) (*buildv1alpha1.Build, error) {
+	return c.bLister.Builds(namespace).Get(name)
+}
+func (c *controller) deleteBuild(namespace, name string) error {
+	return c.bc.BuildV1alpha1().Builds(namespace).Delete(name, &metav1.DeleteOptions{})
+}
+func (c *controller) createBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Create(b)
+}
+func (c *controller) updateBuild(namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	return c.bc.BuildV1alpha1().Builds(namespace).Update(b)
+}
+func (c *controller) updateProwJob(namespace string, pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error) {
+	return c.pjc.ProwV1().ProwJobs(namespace).Update(pj)
+}
+
+func (c *controller) now() metav1.Time {
+	return metav1.Now()
+}
+
+var (
+	groupVersionKind = schema.GroupVersionKind{
+		Group:   prowjobv1.SchemeGroupVersion.Group,
+		Version: prowjobv1.SchemeGroupVersion.Version,
+		Kind:    "ProwJob",
+	}
+)
+
+// reconcile ensures that the prowjob and build
+func reconcile(c reconciler, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource %s: %v", key, err))
+	}
+
+	var wantBuild bool
+
+	pj, err := c.getProwJob(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not want build
+	case err != nil:
+		return fmt.Errorf("get prowjob: %v", err)
+	case pj.Spec.Agent != prowjobv1.BuildAgent:
+		// Do not want a build for this job
+	case pj.DeletionTimestamp == nil:
+		wantBuild = true
+	}
+
+	var haveBuild bool
+
+	b, err := c.getBuild(namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Do not have a build
+	case err != nil:
+		return fmt.Errorf("get build: %v", err)
+	case b.DeletionTimestamp == nil:
```

Comment:
So if we launch a Build, then it finishes and something deletes it, but we catch it after the delete call but before the deletion, we trigger another Build?